app/database/group_group_store.go

```diff
@@ -22,3 +27,53 @@ func (s *GroupGroupStore) WhereUserIsMember(user *User) *DB {
 func (s *GroupGroupStore) createNewAncestors() {
 	s.DataStore.createNewAncestors("groups", "Group")
 }
+
+var ErrRelationAlreadyExists = errors.New("the relation already exists")
+var ErrRelationCycle = errors.New("a group cannot become an ancestor of itself")
+
+const groupsRelationsLockTimeout = 3 * time.Second
+
+func (s *GroupGroupStore) CreateRelation(parentGroupID, childGroupID int64) (err error) {
+	s.mustBeInTransaction()
+	defer recoverPanics(&err)
+
+	mustNotBeError(s.WithNamedLock(s.tableName, groupsRelationsLockTimeout, func(store *DataStore) (err error) {
+		var rows []interface{}
+		mustNotBeError(s.WithWriteLock().
+			Select("ID").
+			Where("idGroupChild = ? AND idGroupParent = ?", childGroupID, parentGroupID).
+			Limit(1).
+			Scan(&rows).Error())
```

Comment:
In MySQL there are no perf difference with a count? (or other constructs)

Reply:
There is a bug in 'gorm'. It skips the write lock ('FOR UPDATE') when counting rows. Probably we should create a PR with the bugfix. It's just 3 lines of code. But without this patch we cannot use Count() here :(